### PR TITLE
[draft] Add metadata support for distributed and cuda memory info

### PR DESCRIPTION
**Pull request by guotuofeng**
Prototype of adding metadata when starting profiler.

### CODE REVIEW

File: libkineto/src/output_json.cpp [hunk 1]
```diff
@@ -31,12 +31,48 @@ namespace KINETO_NAMESPACE {
...
+      }}
+    )JSON";
+  }
+
+  if (!metadata.distributed_.backend_.empty()){
+    stream << fmt::format(R"JSON(
+      "distributed": {{
+        "backend": "{}",
+        "rank": {},
+        "work_size": {}
```
Comment (nbcsm) [inline libkineto/src/output_json.cpp:67]: world

File: libkineto/src/output_json.cpp [hunk 2]
```diff
@@ -31,12 +31,48 @@ namespace KINETO_NAMESPACE {
...
+  
+    for (size_t i = 0; i < metadata.gpus_.size(); i++) {
+      const auto& gpu = metadata.gpus_[i];
+      stream << fmt::format(R"JSON(
+        "GPU {}": {{
+          "id": {},
+          "sku": "{}",
+          "memory": {}
+        }},
+      )JSON", i, i, gpu.name_, gpu.totalMemory_);
```
Comment (nbcsm) [inline libkineto/src/output_json.cpp:54]: `id` must align with `device id`

File: libkineto/src/ActivityProfilerController.cpp [hunk 3]
```diff
@@ -134,6 +134,7 @@ void ActivityProfilerController::prepareTrace(const Config& config) {
   }
 
   profiler_->configure(config, now);
+  profiler_->prepareTrace(metadata);
```
Comment (guyang3532) [inline libkineto/src/ActivityProfilerController.cpp]: Merging these two function seems more concise? like profiler_->configure(config, now, metadata); ?

**Comment by gdankel** [comment]
Thanks for the patch! I will implement it differently for a couple of reasons:
1) We want to support metadata like this but in a more generic way
2) I have some refactoring in progress and I'll add support for this in a cleaner way once that's in place.

**Comment by nbcsm** [comment]
> Thanks for the patch! I will implement it differently for a couple of reasons:
> 
> 1. We want to support metadata like this but in a more generic way
> 2. I have some refactoring in progress and I'll add support for this in a cleaner way once that's in place.

Cool, thanks.
When will this happen?

**Pull request closed without merging**